### PR TITLE
Waterline rework

### PR DESCRIPTION
**Pull request by NotAPenguin0**
# Overall vision

The current waterline is a fairly boring line. Every step has a single input and output (purified water), and there are not really any unique mechanics involved. The goal of this rework is to
- Provide each step of the waterline with a completely unique multiblock, each with a unique automation challenge. These challenges gradually increase in difficulty as you progress through the tiers. 
- Spread out these multis across LuV-UEV progression, so that water purification becomes a process you need to revisit and upgrade as you progress.
- Try to polish these multis as much as I can. Unique textures, lore entries and casings are the norm.
- Avoid adding large *resource costs* associated with operating the waterline. This avoids disturbing game balance too much. The point of most steps will not be to have large passive resource drains, but instead encouraging the player to rely on smart automation designs to avoid these resource costs.
- Avoid any multiblock spam. All waterline multis should only be built once, since they have built in infinite parallel. This is a small widget in the controller that controls the maximum parallel value, similar to what exists in the Transcendent Plasma Mixer. Using these parallels, only *purified water I/O and power* will scale, any catalysts described in tooltips or recipes do not.
- Since these multiblocks scale infinitely, I have decided to remove the recipes for the water QFT catalysts. Existing waterline QFTs will keep working, but since the catalyst is now uncraftable the QFT can no longer be used to skip the waterline in the future.

My most sincere apologies to the person who has to review this chonker of a PR.

# Detailed mechanics

## Water Purification Plant

image
image

Aimed to be tiered at LuV, the water purification plant is the brain of the waterline rework. Its purpose is to serve as a central access point for all other multiblocks in the waterline. It provides power and maintenance fixes to them, and synchronizes the recipe loop. Functionally, this multiblock is not strictly necessary anymore and it can be stripped out of the code, but thematically I like it because it reinforces the player feeling of building a water purification factory. It also serves as additional tooltip space that can explain mechanics shared by all the other multiblocks in the waterline.

This multilbock introduces a couple new casings, shown from left to right below:

image

- *Sterile Water Plant Casing*
- *Reinforced Sterile Water Plant Casing*
- *Superplasticizer-Treated High Strength Concrete*

It also introduces a new type of glass, *Tinted Industrial Glass*, which currently comes in 4 colors:
image

## Step 1: Clarifier

image
image
image

The clarifier should be instantly recognizable to most people who have seen anything related to water treatment before. It's the classic large sifter tank you see in every industrial water purification plant. Tiered at LuV, this is a very gentle introduction to the waterline. The only automation required is refilling the filter items automatically. Doing so requires you to make *Activated Carbon*, which is done using a very short 3-step chemical line.

image
image
image
image

The phosphoric acid input in this line is fully recycled, so the only input is carbon dust and water, making the material cost of this essentially negligible in LuV.

## Step 2: Ozonation

image
image

Ozonation is also tiered at LuV and is similarly simple. The main cost associated with this step is supplying ozone gas, which is a relatively energy intensive laser engraver recipe (**I have buffed the recipe time 10x since posting this because it's not really viable to make this amount in LuV otherwise**). This was done to match how ozone can be obtained in real life. I want to note that this step also has a unique NEI background like the clarifier, but when showing two recipes per page they overlap and look messy. I still need to figure out how to stop this.

image

This multiblock introduces two new casing blocks, shown again from left to right:
image

- *Inert Filtration Casing*
- *Reactive Gas Containment Casing*

## Step 3: Flocculation

image
image
image

In step 3, which will be tiered at ZPM, the first real automation challenge is presented, though it is a fairly simple one. This multiblock requires the player to supply large amounts of a fluid, but in precise amounts. The amount of 100kL was chosen so that it's the largest round number that fits into a ZPM input hatch. This will force the player to be smart about how they insert the fluid, since it cannot be done all at once. Polyaluminium Chloride itself is fairly easy to obtain, and by recycling the output Flocculation Waste Liquid all the chlorine can be recovered so this doesn't end up being an enormous chlorine sink.

image
*Polyaluminium Chloride from Aluminium Hydroxide and Hydrochloric Acid*
image
*Distill Flocculation Waste Liquid into Diluted Hydrochloric Acid, which can then be made into all the HCl needed to make up for the used Polyaluminium Chloride.*

One new casing block is introduced, *Slick Sterile Flocculation Casing*.
image

## Step 4: pH Neutralization

image
image
image
image

Tiered at ZPM and ramping up in complexity, and perhaps my personal favourite, is the pH Neutralization Unit. Naturally inspired by real life processes, the goal is to neutralize the pH value of the water by inserting small amounts of catalysts into it. The amount of catalysts and materials used are not a real cost to consider in ZPM, which is also not the main goal. This will probably require some more intricate redstone or SFM setup to properly automate. The multiblock supports up to two pH Sensor Hatches to read out the internal pH value, which can be used to control the automation.

Three new casing blocks are added for this step, again from left to right:
image
- *Stabilized Naquadah Water Plant Casing*
- *Inert Neutralization Water Plant Casing*
- *Chemical Grade Glass*

## Step 5: Extreme Temperature Fluctuation

image
image
image

Breaking into UV, we can start losing some of the realism and enter the sciencefiction domain a little more. The core principle is in fact fairly similar to step 4, but we add a small layer of complexity here. There's no more hatch to read out the internal value, instead you need to carefully keep track of how much water and plasma has been inserted each recipe cycle to avoid overheating the multiblock.

A single casing block is added, *Heat-Resistant Trinium Plated Casing*
image

## Step 6: High Energy Laser Purification

image
image
(TBA: NEI recipe background with the proper lens order)

Again tiered at UV, this step forces the player to use another redstone hatch, this time indicating when to extract a lens from the input bus and insert a new one. The order of the lenses is fixed and will be shown in NEI, but the correct background assets for this are still missing.

Three new casing blocks are added:
image
- *Naquadria-Reinforced Water Plant Casing*
- *High Energy Ultraviolet Emitter Casing*
- *Electron-Permeable Neutronium Coated Glass*

## Step 7: Residual Decontaminant Degassing

image
image

With a large structure and equally large tooltip, this step of the new waterline will be tiered at UEV. While the tooltip and automation challenge looks very complex, I believe it's not that difficult. There are a lot of instructions, but the individual bits of logic are fairly simple and similar to problems solved in the previous steps (such as inserting exact amounts of fluid). 

One new casing block is added, *Omni-Purpose Infinity Fused Glass*
image

## Step 8: Absolute Baryonic Perfection
image

I will update this one once the final textures and lore entries come in.

# Balancing

Since the waterline will now start much earlier, it also needs to have a use this early. I have opted to change cutting machine and laser engraver recipes so that

- Using **naquadah** doped boules and wafers can optionally take grade 1 or grade 2 purified water, with grade 1 providing a 25% speed bonus and grade 2 providing a 50% speed bonus, This is a large bonus, but hopefully incentivizes actually setting up grade 2 purified water.
- Using **europium** doped boules and wafers will require grade 3 or grade 4 purified water, with grade 4 again providing a 50% speed bonus.
- **Americium** needs grade 5 or grade 6, with the same speed bonus.
- All optical recipes that previously required purified water will now only accept grade 7 or grade 8 purified water. Grade 8 will yield bonus output and provide bonus speed, to further encourage using this.
- To avoid skipping the entire waterline with the beamline, I also added purified water to the recipes for blank photolithographic masks:
    - T1 blank masks use grade 1, 2 or 3 purified water. Using higher than required gives a speed bonus.
    - T2 blank masks use grade 4 or 5 purified water.
    - T3 blank masks use grade 6, 7 or 8 purified water.

Some of these recipe changes are/will be in the coremod PR.

With the infinite parallels and fairly low amounts of water needed, I don't think producing enough water will be a big problem for most bases.

# Disclaimer

Some of the numbers may not be final and can be adjusted after testing in real bases.

### CODE REVIEW

File: src/main/java/gregtech/common/tileentities/machines/multi/purification/GT_MetaTileEntity_PurificationPlant.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
+package gregtech.common.tileentities.machines.multi.purification;
+
+import static com.gtnewhorizon.structurelib.structure.StructureUtility.*;
```
Comment (chochem) [inline src/main/java/gregtech/common/tileentities/machines/multi/purification/GT_MetaTileEntity_PurificationPlant.java:3]: we dont use star imports. you might want to change the settings in your IDE.

Comment (NotAPenguin0) [reply src/main/java/gregtech/common/tileentities/machines/multi/purification/GT_MetaTileEntity_PurificationPlant.java:3]: Will do, I already had to fix it accidentally replacing all imports in the multiblock registry with a single star import.

**Comment by NotAPenguin0** [comment]
I'm putting the final touches on this, adding some more textures and assets to the final step, but since the implementation is mostly done I'll mark this as ready for review. I'll edit the main PR description to include a detailed overview of the entire rework.

**Review by Connor-Colenso — APPROVED** [review]
I have given this a glance over, but I don't realistically see this getting a formal 100% review on 8k lines, but the general direction seems popular and uncontroversial, so lets go with it and iterate on any fixes/balance needed with actual gameplay feedback. 

Really nice work.

**Comment by McFrugal** [comment]
Concerns:
1, water purification plants are in place in real life because they're more efficient than distillation for drinking water and industrial purposes.  Since energy for distillation is basically free in GTNH due to it being a low level recipe, this driving force doesn't apply anymore.  Which then means, at no step of this design should you be thinking about what real life processes require in energy cost or whatever.  That lesson should have been clear from the required Ozone buff so maybe you understand already.
2, we have access to recipes that simply react Hydrogen and Oxygen to get water.  Why is this not sufficiently pure anymore, despite being perfectly chemically pure?

Suggestion:
Since IRL Clarifiers are for removing sediment contaminants such as those in sewage, not purifying already-drinkable water, I'd suggest this machine be entirely skipped.

**Comment by FourIsTheNumber** [comment]
Purified water line already exists in GTNH, it's just bad and uninteresting, so 2 seems kind of irrelevant. The goal is obviously not to make a realistic water plant simulator so this is all moot.

**Comment by McFrugal** [comment]
> Purified water line already exists in GTNH, it's just bad and uninteresting, so 2 seems kind of irrelevant. The goal is obviously not to make a realistic water plant simulator so this is all moot.

Right, but it's water.  It doesn't need to be interesting, it's one of the simplest and most common chemical compounds in existence.

**Comment by FourIsTheNumber** [comment]
> > Purified water line already exists in GTNH, it's just bad and uninteresting, so 2 seems kind of irrelevant. The goal is obviously not to make a realistic water plant simulator so this is all moot.
> 
> Right, but it's water. It doesn't need to be interesting, it's one of the simplest and most common chemical compounds in existence.

Who cares? This is a fun line for a video game that's mimicking a real world process. Nothing about GTNH is hyper-realistic so trying to nitpick a cool and interesting addition to the pack because it isn't realistic enough for your tastes is genuinely bizarre.

**Comment by McFrugal** [comment]
I'll retract concern 2 on the grounds of purified water grades already being a thing that I didn't know about.

Clarifier could be neat for cleaning polluted water resulting from high levels of pollution... or maybe methods of producing large quantities of water (reservoir?) creates polluted/silty water that you have to process with it to get regular water.

**Comment by Krispion** [comment]
Pretty much agree with McFrugal, why so much effort (and so many new bespoke multis - ton of bloat, and they probably have 1-2 recipes each) just for purifying water? I mean it is what it is, but why not spend this effort on something interesting...

**Comment by FourIsTheNumber** [comment]
> Pretty much agree with McFrugal, why so much effort (and so many new bespoke multis - ton of bloat, and they probably have 1-2 recipes each) just for purifying water? I mean it is what it is, but why not spend this effort on something interesting...

New mechanics and systems aren't bloat. The current waterline is pretty hated, and imo this line is plenty interesting

**Comment by LazyFleshWasTaken** [comment]
> Pretty much agree with McFrugal, why so much effort (and so many new bespoke multis - ton of bloat, and they probably have 1-2 recipes each) just for purifying water? I mean it is what it is, but why not spend this effort on something interesting...

This IS interesting, and will have actual uses in progression. The current waterline is basic as hell and boring too, an upgrade on it is welcome any time.

**Comment by NotAPenguin0** [comment]
> why so much effort (and so many new bespoke multis - ton of bloat, and they probably have 1-2 recipes each) just for purifying water?

Very simply put, because I enjoyed doing it and because I don't see a major negative impact. 

To answer earlier concerns, of course it's not extremely realistic and that's not the goal either. I've always put gameplay above realism even if there is some inspiration. Yes, in GTNH we have access to pure hydrogen and oxygen (though you could argue that's not exactly realistic either), but in the end you shouldn't be overthinking the realism part and try to just enjoy the game as it is.

**Comment by Hikari1nVoid** [comment]
Thank you for the effort in water line Rev2.
However Why did you nuke current QFT recipes when G8W multi is incomplete and hatches for G7W is missing recipe?

**Comment by ChemEng227** [comment]
HOW can I even make a possible automatation to these machines, Especially the len swapping of T6, and these signal translation for T7?
The only possible way I can come up with is using Opencomputer, but that will be too unfriendly for those who cannot program lua.

**Comment by NotAPenguin0** [comment]
> Thank you for the effort in water line Rev2.
> 
> However Why did you nuke current QFT recipes when G8W multi is incomplete and hatches for G7W is missing recipe? 

I wasn't aware the hatch was missing recipes, so those will be added as soon as possible. Sorry for the inconvenience, but this stuff happens with nightlies sometimes.

**Comment by NotAPenguin0** [comment]
> HOW can I even make a possible automatation to these machines, Especially the len swapping of T6, and these signal translation for T7?
> 
> Even SFM seemed not powerful enough for these conplicated ways. The only possible way I can come up with is using Opencomputer, but that will be too unfriendly for those who cannot program lua.
> 
> Anyway I dont want to spend all my playing time in front of a unautomatable machine...

Both are possible with pure redstone, made a lot easier with ProjectRed logic gates and programmable circuits.

**Comment by ChemEng227** [comment]
> > HOW can I even make a possible automatation to these machines, Especially the len swapping of T6, and these signal translation for T7?
> > Even SFM seemed not powerful enough for these conplicated ways. The only possible way I can come up with is using Opencomputer, but that will be too unfriendly for those who cannot program lua.
> > Anyway I dont want to spend all my playing time in front of a unautomatable machine...
> 
> Both are possible with pure redstone, made a lot easier with ProjectRed logic gates and programmable circuits.

If so that will be better than I thought... But these complicated mechanics seemed to be easily leading to a situation where players just copy IC blueprints around without making their own solution. That's not a good gameplay desingning in my mind, though.

**Comment by NotAPenguin0** [comment]
> > > HOW can I even make a possible automatation to these machines, Especially the len swapping of T6, and these signal translation for T7?
> > > Even SFM seemed not powerful enough for these conplicated ways. The only possible way I can come up with is using Opencomputer, but that will be too unfriendly for those who cannot program lua.
> > > Anyway I dont want to spend all my playing time in front of a unautomatable machine...
> > 
> > 
> > Both are possible with pure redstone, made a lot easier with ProjectRed logic gates and programmable circuits.
> 
> If so that will be better than I thought... But these complicated mechanics seemed to be easily leading to a situation where players just copy IC blueprints around without making their own solution. That's not a good gameplay desingning in my mind, though.

This is an issue in any automation game. People can play factorio by copying their entire base from online blueprints, that doesn't make it a bad game or doesn't meant the automation puzzles in it are not worth having. Players who want to make their own solution get the opportunity to, and players who just want to copy things without using their brain can go look up the solution.

**Comment by ChemEng227** [comment]
> This is an issue in any automation game. People can play factorio by copying their entire base from online blueprints, that doesn't make it a bad game or doesn't meant the automation puzzles in it are not worth having. Players who want to make their own solution get the opportunity to, and players who just want to copy things without using their brain can go look up the solution.

Yes, you are right here. I shall have explained my opinions more clearly.
What I actually wanted to express here is that the mechanics, especially T7, had a understanding cost higher than an average player expected (Imagine someone reading the tooltip and says "Oh s*** what is binary"). That will make them braindead and turn to nothing but copying.

**Comment by RealSilverMoon** [comment]
Partially agree with ChemEng227 . These mechanics are overly complex **relative to the original GTNH content**, which would drastically increase the skill requirements for the normal players. Due to the mandatory requirements for wafer fabrication, some players may even **abandon** the pack for this reason. Furthermore, such a large and automation-challenging mechanism is only used to produce "purified water" ( and yes, I do wonder why "quark creation catalysts" are needed for super-purified water that can be easily synthesized with hydrogen and oxygen in the LCR. That defies logic even in a game), the benefits are disproportionate to the effort, which doesn't give much feedback to the player - at least not the joy of making a set of EOH automations.
You might consider:

- Setting up a simplified mechanic, but with reduced production (instead of just zeroing it out). This gives players looking for a challenge the option to work on more complex mechanisms.

- Adding an AI controller, perhaps craftable at three to four voltage levels above this module, and using the AI controller will instantly simplify the mechanics and maximize yield.

- Change the name. At least not "Level 8 Pure Water" but something like "Perfection Baryonic Liquid".

**Comment by FourIsTheNumber** [comment]
> Partially agree with ChemEng227 . These mechanics are overly complex **relative to the original GTNH content**, which would drastically increase the skill requirements for the normal players. Due to the mandatory requirements for wafer fabrication, some players may even **abandon** the pack for this reason. Furthermore, such a large and automation-challenging mechanism is only used to produce "purified water" ( and yes, I do wonder why "quark creation catalysts" are needed for super-purified water that can be easily synthesized with hydrogen and oxygen in the LCR. That defies logic even in a game), the benefits are disproportionate to the effort, which doesn't give much feedback to the player - at least not the joy of making a set of EOH automations. You might consider:
> 
>     * Setting up a simplified mechanic, but with reduced production (instead of just zeroing it out). This gives players looking for a challenge the option to work on more complex mechanisms.
> 
>     * Adding an AI controller, perhaps craftable at three to four voltage levels above this module, and using the AI controller will instantly simplify the mechanics and maximize yield.
> 
>     * Change the name. At least not "Level 8 Pure Water" but something like "Perfection Baryonic Liquid".

No one is going to quit the pack over having to do an automation challenge in an automation modpack at ZPM+. At worst they will come to discord and ask for the optimal setup.

**Comment by ChemEng227** [comment]
> No one is going to quit the pack over having to do an automation challenge in an automation modpack at ZPM+. At worst they will come to discord and ask for the optimal setup.

For this part I'll agree with RealSilverMoon. The majority of current GTNH automation challange is about building production chains and putting correct things into correct machines with correct energy level, e.g. Pt line and Kevlar. There's only few machines with extra mechanics. The most complicated of them is just EOH, but even EOH mechanics is still clear for understanding and automation (by AE fake crafting), unlike this T7 water, which forces one to grab a notebook and do some logical calculation before understanding it. It's not good to force a large percentage of players go to the community for help, in my opinion.

**Comment by FourIsTheNumber** [comment]
> > No one is going to quit the pack over having to do an automation challenge in an automation modpack at ZPM+. At worst they will come to discord and ask for the optimal setup.
> 
> For this part I'll agree with RealSilverMoon. The majority of current GTNH automation challange is about building production chains and putting correct things into correct machines with correct energy level, e.g. Pt line and Kevlar. There's only few machines with extra mechanics. The most complicated of them is just EOH, but even EOH mechanics is still clear for understanding and automation (by AE fake crafting), unlike this T7 water, which forces one to grab a notebook and do some logical calculation before understanding it. It's not good to force a large percentage of players go to the community for help, in my opinion.

This is nonsense. The vast majority of players I've spoken to want GTNH to have more automation challenges (and this is the view of the development team at large, as far as I'm aware). If you want to have automation challenges you have to start adding them somewhere. These mechanics are not so complicated that you have to ask for help - players will simply do it anyways. Which is fine, it's what is already done for the vast majority of mechanics.

Consider things like assembling line, which ask you to do precise insertion of components in particular slots without overfilling, sometimes splitting stacks. This is a classic GT automation puzzle, you just don't think of it because it has been long solved by the community and most players use the same copy-pasted solution. Consider other mechanics like bacterial vat or looping logistical issues in processing lines. 

Bottom line: This is an automation modpack. It should have logistical/automation challenges, that is literally the entire point and it is what most players are attracted to in the first place. Cutting out mechanical depth because a few players will whine and ask for answers in discord is entirely unacceptable.

**Comment by NotAPenguin0** [comment]
I fully agree with what Four said. As a side note, if you'd think about the multiblocks more carefully you'd realize there are (very wasteful) ways to set them up with minimal automation for ~10-20% the intended production rate.

**Comment by Pxx500** [comment]
We need more mechanics and more variety, this is what makes the game interesting. This is definitely not too hard considering that most casual players already give up before LuV, at tiers this high we can do crazy automation tasks and not worry about people leaving the pack.
